src/hotspot/cpu/x86/x86_64.ad

```diff
@@ -6779,8 +7310,22 @@ instruct addL_rReg(rRegL dst, rRegL src, rFlagsReg cr)
   ins_pipe(ialu_reg_reg);
 %}
 
+instruct addL_rReg_ndd(rRegL dst, rRegL src1, rRegL src2, rFlagsReg cr)
+%{
+  predicate(UseAPX);
+  match(Set dst (AddL src1 src2));
+  effect(KILL cr);
+
+  format %{ "eaddq    $dst, $src1, $src2\t# long ndd" %}
+  ins_encode %{
+    __ eaddq($dst$$Register, $src1$$Register, $src2$$Register, false);
+  %}
```

Comment:
Current scheme is favoring emission of NDD instruction on APX enabled targets, even if destination and source registers belongs legacy GPR set. We should extend assembler layer to demote EEVEX to REX encoding if dst matches with source operands.

FYI, processor backend also uses Move Elimination to prevent dispatch of  GPR to GPR moves to execution ports.

This can be used to further break NDD pattens with all different legacy register operands, GPR to GPR moves should consume no more than 3 bytes encoding 1 byte for REX prefix, 1 for opcode and 1 for ModRM byte, still less than 4 byte prefix.